add ptz to UI and switches. Missing docs and testing

dhruvb14 opened the pull request:
# Features
- Add Homekit toggle to allow moving PTZ camera's to Unifi Presets

# Other
- Fix comment typos

dhruvb14 commented:
@hjdhjd I commented on your [other repo](https://github.com/hjdhjd/unifi-protect/issues/41#issuecomment-2495673626) and really appreciate you pointing me in the correct direction. I have taken my first attempt at adding this feature to HBUP but am still feeling like I am missing something. I have yet to test this out yet (working on setting up a dev HB instance) but would love some early feedback on the PR as I'm just reading similar implementations in your existing code to try and implement this feature.

Am I on the right track or did I miss something major?

hjdhjd commented:
Appreciate the effort at adding this, but to be honest, it has a long way to go before it’s going to get accepted. I’m not going to provide line-by-line feedback…I simply do not have the time commitment to do so, but briefly:

- On the smaller end - your formatting, both code and grammatical in your comments, are inconsistent. Comments should be sentences. Sentences end in periods. Capitalization matters to me. Code and comments should be beautiful and expositive. Any PR must fully pass all linting without any warnings.
- On the medium end - absolutely should not be trying to save anything to the context of the accessory under any circumstances that might be encountered for this feature. Period.
- On the larger end - adding multiple feature options is a non-starter. You expose PTZ features or you don’t, and that’s it. You should be able to gracefully handle an infinite number of presets - locking in the number of presets artificially isn’t a great approach for longevity.
- On the larger end - you must autodetect PTZ capabilities…adding module-level variables should only be done when absolutely necessary for some larger purpose. This feature doesn’t meet that bar.
- On the larger end - the logic and flow of the code itself needs to be clear, consistent, and elegant. Doing things like `const featureName = "Device.Ptz.Preset.Switch." + ptzPresetSwitch; if(!this.hasFeature(featureName)) { …` is an example of what I am referencing.

I hope you take the feedback in the spirit it’s intended. It’s got a some distance to go, but if you keep at it, happy to review it at various milestones along the way and eventually consider it for inclusion.

dhruvb14 commented:
@hjdhjd I really appreciate your feedback! I fully understand this codebase is well written and very thorough. I do not wish to tarnish it by lowering your code quality. I did not take any hard feelings what so ever to your feedback and have pushed a few updates. I also had a few questions on how you propose a workaround since Unifi doesn't allow implementing some of your recommendations.

```
On the smaller end - your formatting, both code and grammatical in your comments, are inconsistent. Comments should be sentences. Sentences end in periods. Capitalization matters to me. Code and comments should be beautiful and expositive. Any PR must fully pass all linting without any warnings.
```
Understood! I will work on cleaning this up! I have eslinted my code and everything was passing locally. I was on a long car ride and wrote all this code during the ride so was more focused on understanding functionality! I'll get the naming all cleaned up along with better comments.

```
On the medium end - absolutely should not be trying to save anything to the context of the accessory under any circumstances that might be encountered for this feature. Period.
```
I was not aware I did this, I was just copying the implementation of the NVR Recording Switch implementation since its similar to what I was trying to achieve. I'll keep looking but since the entire endpoint is stateless (no way to see what preset the camera is located at via Unifi's API or how many presets the user has configured) I am completely stateless. I made a few updates since your comment and removed some code to clean it up. When you activate a preset, I have it setup to disable the switch after 1 second so it's similar to a button (which HomeKit doesn't allow you to push a button in HomeKit app so a switch which returns to off is how I implemented this)

```
On the larger end - adding multiple feature options is a non-starter. You expose PTZ features or you don’t, and that’s it. You should be able to gracefully handle an infinite number of presets - locking in the number of presets artificially isn’t a great approach for longevity.
```
This is ugly but the only way to do it from what I understand. Unifi does NOT expose any information about the Presets (I made a rest call to the camera's get endpoint and it does not have any information about PTZ positioning or presets), Unifi allow a max of 9 preset and one home position. You do not have to create all 9 presets so my intention was to expose only the ones you need for your device. In my example I have 4 presets and home preset exposed on my camera. All the other presets do not return an error from the rest endpoint, they just do nothing. 

I am severely limited by Unifi's implementation and would love to brainstorm how to work around this.

```
On the larger end - you must autodetect PTZ capabilities…adding module-level variables should only be done when absolutely necessary for some larger purpose. This feature doesn’t meet that bar.
```

I do detect PTZ capabilities and only show the features if the camera supports ptz. I validated this by exposing one ptz camera and one non-ptz camera and the features only show on the ptz cameras.

<img width="781" alt="Screenshot 2024-11-27 at 6 18 06 PM" src="https://github.com/user-attachments/assets/f64b2152-a32e-4573-b053-6fc1aeeb4740">
<img width="779" alt="Screenshot 2024-11-27 at 6 17 55 PM" src="https://github.com/user-attachments/assets/e1ffeea2-fb0f-44a9-a21f-1eeff9bab83e">

```
On the larger end - the logic and flow of the code itself needs to be clear, consistent, and elegant. Doing things like const featureName = "Device.Ptz.Preset.Switch." + ptzPresetSwitch; if(!this.hasFeature(featureName)) { … is an example of what I am referencing.

```

This was a direct copy of how the existing code for `configureNvrRecordingSwitch` ([code link](https://github.com/hjdhjd/homebridge-unifi-protect/blob/main/src/devices/protect-camera.ts#L1125)) which does exactly this so I was under the impression this was the convention for implementation. `eslint` said I could not do string interpolation so I had to refactor to using a variable instead for the dynamic nature of the ptz rest call.


If you could reply to some of the Unifi API limitations and how you want me to proceed ill code it all up. I have another 5 hour car ride Saturday so I should have a bunch of time to look into this.

dhruvb14 commented:
Checked Linting:
<img width="841" alt="Screenshot 2024-11-27 at 8 35 25 PM" src="https://github.com/user-attachments/assets/6d32ab1a-1e62-43fb-9e72-8526df99cdce">

dhruvb14 commented:
Converting back to `draft` as I found the endpoint `/ptz/preset` which returns all configured PTZ presets so I can dynamically put this behind 1 device checkbox as recommended by @hjdhjd. Working on implementing the code changes.
